README.md
# Ring Size View version 1.0.8

A Flutter package that provides a custom widget for calculating and displaying ring sizes.

## Demo new


<img src="https://github.com/PuzzleTakX/ring_size_view/blob/master/demo/demo.jpg?raw=true" alt="image_demo" width="300" height="700">


## Installation

To use this package, add `ring_size_view` as a dependency in your `pubspec.yaml` file.

```yaml
dependencies:
  flutter:
    sdk: flutter
  ring_size_view: ^1.0.8 # Use the latest version from pub.dev
  ````

Usage

````import 'package:ring_size_view/ring_size_view.dart';````


## RingSizeView Widget

The `RingSizeView` widget is a custom Flutter widget provided by the `ring_size_view` package. It is designed to calculate and display the size information of a ring based on the user's input.

### Properties

- `ringSize`: A double value representing the current ring size. This value is used to calculate the diameter, circumference, and radius of the ring. You can bind this value to a slider or any other input widget to allow the user to adjust the ring size.

- `calculateDiameter(value)`: A callback function that is called when the ring size is adjusted by the user. It receives the calculated diameter value as an argument. You can use this function to update the diameter value in your app.

- `calculateRadius(value)`: A callback function that is called when the ring size is adjusted. It receives the calculated radius value as an argument. You can use this function to perform any additional actions based on the radius value.

- `calculateCircumference(value)`: A callback function that is called when the ring size is adjusted. It receives the calculated circumference value as an argument. You can use this function to update the circumference value in your app.

- `height`: A double value representing the height of the `RingSizeView` widget.

- `width`: A double value representing the width of the `RingSizeView` widget.

### Example Usage

```dart
RingSizeView(
ringSize: ringSize,
calculateDiameter: calculateDiameter,
calculateRadius: calculateRadius,
calculateCircumference: calculateCircumference,
height: 190,
width: 190,
)
````


## Calculation Functions

The `calculateCircumference`, `calculateRadius`, and `calculateDiameter` functions are used in the `RingSizeView` widget to perform calculations and update the corresponding values.

### `calculateCircumference(value)`

This function is called when the user interacts with the `RingSizeView` widget and updates the value of `circumference`. It takes the calculated circumference value as an argument and updates the `circumference` variable. For example:

```dart
calculateCircumference(double value) {
  circumference = value;
  print("Info calculateCircumference : $value");
}
````

* ring size
* ring size view
* flutter
* puzzletak

## Tags

<meta name="keywords" content="ring size, ring size view, flutter, puzzletak">

Features

    Calculate ring diameter, circumference, and radius.
    Slider to adjust the ring size.
    Customizable styles.

Getting Started

For more details and examples, check out the documentation.
Contributing

Contributions are welcome! Please fork the repository and submit a pull request.
License

This project is licensed under the Apache License 2.0 - see the LICENSE file for details.


CHANGELOG.md
# Ring Size View version 1.0.8

A Flutter package that provides a custom widget for calculating and displaying ring sizes.

## Demo new

<img src="https://github.com/PuzzleTakX/ring_size_view/blob/master/demo/demo.jpg?raw=true" alt="image_demo" width="300" height="700">



## Installation

To use this package, add `ring_size_view` as a dependency in your `pubspec.yaml` file.

```yaml
dependencies:
  flutter:
    sdk: flutter
  ring_size_view: ^1.0.8 # Use the latest version from pub.dev
  ````

Usage

````import 'package:ring_size_view/ring_size_view.dart';````


## RingSizeView Widget

The `RingSizeView` widget is a custom Flutter widget provided by the `ring_size_view` package. It is designed to calculate and display the size information of a ring based on the user's input.

### Properties

- `ringSize`: A double value representing the current ring size. This value is used to calculate the diameter, circumference, and radius of the ring. You can bind this value to a slider or any other input widget to allow the user to adjust the ring size.

- `calculateDiameter(value)`: A callback function that is called when the ring size is adjusted by the user. It receives the calculated diameter value as an argument. You can use this function to update the diameter value in your app.

- `calculateRadius(value)`: A callback function that is called when the ring size is adjusted. It receives the calculated radius value as an argument. You can use this function to perform any additional actions based on the radius value.

- `calculateCircumference(value)`: A callback function that is called when the ring size is adjusted. It receives the calculated circumference value as an argument. You can use this function to update the circumference value in your app.

- `height`: A double value representing the height of the `RingSizeView` widget.

- `width`: A double value representing the width of the `RingSizeView` widget.

### Example Usage

```dart
RingSizeView(
ringSize: ringSize,
calculateDiameter: calculateDiameter,
calculateRadius: calculateRadius,
calculateCircumference: calculateCircumference,
height: 190,
width: 190,
)
````


## Calculation Functions

The `calculateCircumference`, `calculateRadius`, and `calculateDiameter` functions are used in the `RingSizeView` widget to perform calculations and update the corresponding values.

### `calculateCircumference(value)`

This function is called when the user interacts with the `RingSizeView` widget and updates the value of `circumference`. It takes the calculated circumference value as an argument and updates the `circumference` variable. For example:

```dart
calculateCircumference(double value) {
  circumference = value;
  print("Info calculateCircumference : $value");
}
````

* ring size
* ring size view
* flutter
* puzzletak

## Tags

<meta name="keywords" content="ring size, ring size view, flutter, puzzletak">

Features

    Calculate ring diameter, circumference, and radius.
    Slider to adjust the ring size.
    Customizable styles.

Getting Started

For more details and examples, check out the documentation.
Contributing

Contributions are welcome! Please fork the repository and submit a pull request.
License

This project is licensed under the Apache License 2.0 - see the LICENSE file for details.

test/ring_size_view_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:ring_size_view/ring_size_view.dart';

void main() {
  test('adds one to input values', () {
    return RingSizeView(ringSize: 6.5, calculateDiameter: (e){},
        calculateRadius: (e){},
        calculateCircumference: (e){},
        height: 190, width: 190);
  });
}


lib/ring_size_view.dart
library ring_size_view;

import 'dart:math';

import 'package:flutter/material.dart';



class RingSizeView extends StatefulWidget {
  final Color? backgroundColor;
  final TextDirection? textDirection;
  final TextStyle? textStyle;
  final Function(double) calculateDiameter;
  final Function(double) calculateRadius;
  final Function(double) calculateCircumference;
  final double ringSize;
  final double width;
  final double height;
  const RingSizeView({super.key,
     this.backgroundColor,
     this.textDirection,
    required this.ringSize,
     this.textStyle,
    required this.calculateDiameter,
    required this.calculateRadius,
    required this.calculateCircumference,
    required this.height,
    required this.width,
  });

  @override
  State<RingSizeView> createState() => _RingSizeViewState();
}


class _RingSizeViewState extends State<RingSizeView> {
  ScrollController controller = ScrollController();

  Color backgroundColor = Colors.grey;
  TextDirection textDirection = TextDirection.ltr;
  TextStyle textStyle = const TextStyle(color:  Colors.black,fontSize: 12);


  init(){
    if(widget.backgroundColor != null) backgroundColor = widget.backgroundColor!;
    if(widget.textDirection != null) textDirection = widget.textDirection!;
    if(widget.textStyle != null) textStyle = widget.textStyle!;

  }

  @override
  void initState() {
    init();
    super.initState();
  }

  double calculateDiameter(double size) {
    // Calculate the diameter based on ring size (in millimeters)
    widget.calculateDiameter.call(size * 2.0);
    widget.calculateRadius.call(size);
    widget.calculateCircumference.call(size * pi);
    return size * 2.0;

  }

  double calculateRadius(double size) {
    // Calculate the radius based on ring size (in millimeters)

    return size;
  }

  double calculateCircumference(double size) {
    // Calculate the circumference based on ring size (in millimeters)

    return size * pi;
  }



  @override
  Widget build(BuildContext context) {
    double diameter = calculateDiameter(widget.ringSize);

    return Container(width: widget.width,height: widget.height,
      color:  Colors.blueGrey.withAlpha(10),
      child: Center(
        child: Stack(
          children: [
            SizedBox(
              width: double.maxFinite,
              height: double.maxFinite,
              child: GridView.builder(
                itemCount: 1000,
                shrinkWrap: true,
                physics: const NeverScrollableScrollPhysics(),
                controller: ScrollController(initialScrollOffset: 30),
                gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
                    crossAxisCount: 15,
                    crossAxisSpacing: 0.0,
                    mainAxisSpacing: 0.0
                ),
                itemBuilder: (BuildContext context, int index) {
                  return Container(
                    decoration: BoxDecoration(
                      border: Border.all(color: Colors.black.withAlpha(20),width: 0.3),
                    ),

                  );
                },
              ),
            ),

            Column(
              children: [

                const Spacer(),
                Container(width: double.maxFinite,height: 0.9,color: backgroundColor,),
                SizedBox(
                  width:  (diameter * (6.100)),
                  height:  (diameter * (6.100)),
                ),
                Container(width: double.maxFinite,height: 0.9,color: backgroundColor,),
                const Spacer(),
              ],
            ),
            Row(
              children: [

                const Spacer(),
                Container(width: 0.9,height: double.maxFinite,color: backgroundColor,),
                SizedBox(
                  width:  (diameter * (6.100)),
                  height:  (diameter * (6.100)),
                ),
                Container(width: 0.9,height: double.maxFinite,color: backgroundColor,),
                const Spacer(),
              ],
            ),
            Center(child: Container(
              width:  (diameter * (6.100)),
              height:  (diameter * (6.100)),
              decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  color: Colors.grey.withAlpha(40),
                  border: Border.all(color: Colors.blueGrey.withAlpha(180),width: 2.5)
              ),
              child: Row(
                children: [
                  Icon(Icons.arrow_back_rounded,size: 15,color: backgroundColor,),
                  Flexible(child: Container(width: double.maxFinite,color: backgroundColor,height: 1)),
                  Container(decoration: BoxDecoration(
                    color: backgroundColor.withAlpha(150),
                    borderRadius: BorderRadius.circular(2),
                  ),
                    padding: const EdgeInsets.only(top: 2,bottom: 2,left: 5,right: 5),
                    child: Text('${diameter.toStringAsFixed(2)} mm',
                      textDirection: textDirection,
                      style: textStyle.copyWith(
                        fontSize: (diameter / 2.4)
                      ),
                    ),
                  ),

                  Flexible(child: Container(width: double.maxFinite,color: backgroundColor,height: 1)),
                  Icon(Icons.arrow_forward_rounded,size: 15,color: backgroundColor,),

                ],
              ),
            ),)
          ],
        ),
      ),
    );
  }
}


lib/model_gold_size.dart
class ModelGoldSize {
  DataGoldSize? data;

  ModelGoldSize({this.data});

  ModelGoldSize.fromJson(Map<String, dynamic> json) {
    data = json['data'] != null ? DataGoldSize.fromJson(json['data']) : null;
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    if (this.data != null) {
      data['data'] = this.data!.toJson();
    }
    return data;
  }
}

class DataGoldSize {
  List<GoldSizes>? sizes;

  DataGoldSize({this.sizes});

  DataGoldSize.fromJson(Map<String, dynamic> json) {
    if (json['sizes'] != null) {
      sizes = <GoldSizes>[];
      json['sizes'].forEach((v) {
        sizes!.add(GoldSizes.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    if (sizes != null) {
      data['sizes'] = sizes!.map((v) => v.toJson()).toList();
    }
    return data;
  }
}

class GoldSizes {
  String? diameter;
  String? usa;
  String? australia;
  String? europe;
  String? japan;

  GoldSizes({this.diameter, this.usa, this.australia, this.europe, this.japan});

  GoldSizes.fromJson(Map<String, dynamic> json) {
    diameter = json['diameter'];
    usa = json['usa'];
    australia = json['australia'];
    europe = json['europe'];
    japan = json['japan'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    data['diameter'] = diameter;
    data['usa'] = usa;
    data['australia'] = australia;
    data['europe'] = europe;
    data['japan'] = japan;
    return data;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:ring_size_view/list_rings_sizer_country.dart';
import 'package:ring_size_view/model_gold_size.dart';
import 'package:ring_size_view/ring_size_view.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'RingSizeView',
      theme: ThemeData(

        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'PuzzleTak RingSizeView'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
   double ringSize = 6.5;
   double diameter = 0.0;
   double circumference = 0.0;
   ModelGoldSize modelGoldSize = ModelGoldSize();


   @override
  void initState() {
     init();
    super.initState();
  }


  init(){
     modelGoldSize = ListRingsSizerCountry.getListSings();
  }


  @override
  Widget build(BuildContext context) {


    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Column(

        children: [

          Container(
            padding: const EdgeInsets.only(left: 10,right: 10),
            width: double.maxFinite,height: 180,
            child: Row(
              textDirection: TextDirection.ltr,
              children: [
                RingSizeView(
                    ringSize: ringSize,
                    calculateDiameter: calculateDiameter,
                    calculateRadius: calculateRadius,
                    calculateCircumference: calculateCircumference,
                    height: 190, width: 190),
                Expanded(child: Container(
                  padding: const EdgeInsets.only(left: 10,right: 10,top: 8,bottom: 8),
                  child: Column(
                    children: [
                      Row(
                        textDirection: TextDirection.ltr,
                        children: [
                          const Text(
                            'radius',
                            textDirection: TextDirection.ltr,
                            style: TextStyle(fontSize: 12),
                          ),
                          const SizedBox(width: 3,),
                          Expanded(child: Container(height: 1,color: Colors.deepPurple.withAlpha(150),)),
                          const SizedBox(width: 3,),
                          Text(
                            '${ringSize.toStringAsFixed(2)} mm',
                            textDirection: TextDirection.ltr,
                            style: const TextStyle(fontSize: 12),
                          ),
                        ],
                      ),
                      const SizedBox(height: 15,),

                      Row(
                        textDirection: TextDirection.ltr,
                        children: [
                          const Text(
                            'diameter',
                            textDirection: TextDirection.ltr,
                            style: TextStyle(fontSize: 12),
                          ),
                          const SizedBox(width: 3,),
                          Expanded(child: Container(height: 1,color: Colors.deepPurple.withAlpha(150),)),
                          const SizedBox(width: 3,),
                          Text(
                            '${diameter.toStringAsFixed(2)} mm',
                            textDirection: TextDirection.ltr,
                            style: const TextStyle(fontSize: 12),
                          ),
                        ],
                      ),
                      const SizedBox(height: 15,),
                      Row(
                        textDirection: TextDirection.ltr,
                        children: [
                          const Text(
                            'circumf..',
                            textDirection: TextDirection.ltr,
                            style: TextStyle(fontSize: 12),
                          ),
                          const SizedBox(width: 3,),
                          Expanded(child: Container(height: 1,color: Colors.deepPurple.withAlpha(150),)),
                          const SizedBox(width: 3,),
                          Text(
                            '${circumference.toStringAsFixed(2)} mm',
                            textDirection: TextDirection.ltr,
                            style: const TextStyle(fontSize: 12),
                          ),
                        ],
                      ),
                      const SizedBox(height: 15,),
                      SizedBox(
                        width: double.maxFinite,
                        height: 35,
                        child: TextButton(
                            style: TextButton.styleFrom(
                              backgroundColor: Colors.deepPurple.withAlpha(40)
                            ),
                            onPressed: () async {
                              await Clipboard.setData(ClipboardData(text: ""
                                  " circumference : ${circumference.toStringAsFixed(2)} mm"
                                  "\n"
                                  " diameter : ${diameter.toStringAsFixed(2)} mm"
                                  "\n"
                                  " radius : ${ringSize.toStringAsFixed(2)} mm"
                                  ""));
                            }, child: Text("Copy",
                        textDirection: TextDirection.ltr,
                          style: const TextStyle(fontSize: 12).copyWith(
                            color: Colors.deepPurple,
                            fontWeight: FontWeight.bold

                          ),
                        )),
                      ),

                    ],
                  ),
                ))
              ],
            ),
          ),
          Slider(
            value: ringSize,
            min: 3.5,
            max: 13.0,
            onChanged: (double value) {
              setState(() {
                ringSize = value;
              });
            },
          ),
          Container(width: double.maxFinite,
          height: 40,
            decoration: BoxDecoration(
              color: Colors.grey.withAlpha(40)
            ),
            child: Row(
              children: [
                Flexible(flex: 1,child: Container(
                  alignment: Alignment.center,
                  child: const Text("diameter",
                    textDirection: TextDirection.ltr,
                    style: TextStyle(fontSize: 12),
                  ),
                ),),
                Flexible(flex: 1,child: Container(
                  alignment: Alignment.center,
                  child: const Text("America",
                    textDirection: TextDirection.ltr,
                    style: TextStyle(fontSize: 12),
                  ),
                ),),
                Flexible(flex: 1,child: Container(
                  alignment: Alignment.center,
                  child: const Text("Japan",
                    textDirection: TextDirection.ltr,
                    style: TextStyle(fontSize: 12),
                  ),
                ),),
                Flexible(flex: 1,child: Container(
                  alignment: Alignment.center,
                  child: const Text("Europe",
                    textDirection: TextDirection.ltr,
                    style: TextStyle(fontSize: 12),
                  ),
                ),),
              ],
            ),
          ),

          Expanded(child: ListView.builder(
            itemCount: modelGoldSize.data!.sizes!.length,
          itemBuilder: (context, index) {
            return Padding(
              padding: const EdgeInsets.only(top: 1,bottom: 1),
              child: Container(width: double.maxFinite,
                height: 40,

                decoration: BoxDecoration(
                    color: (ringSize == double.parse(modelGoldSize.data!.sizes![index].diameter!) / 2) ? Colors.deepPurple.withAlpha(40) : Colors.grey.withAlpha(20)
                ),
                child: TextButton(
                  onPressed: (){
                    setState(() {
                      ringSize = double.parse(modelGoldSize.data!.sizes![index].diameter!) / 2;
                    });
                  },
                  child: Row(
                    children: [
                      Flexible(flex: 1,child: Container(
                        alignment: Alignment.center,
                        child: Text(modelGoldSize.data!.sizes![index].diameter!,
                          textDirection: TextDirection.ltr,
                          style: const TextStyle(fontSize: 12).copyWith(

                          ),
                        ),
                      ),),
                      Flexible(flex: 1,child: Container(
                        alignment: Alignment.center,
                        child: Text(modelGoldSize.data!.sizes![index].usa!,
                          textDirection: TextDirection.ltr,
                          style: const TextStyle(fontSize: 12).copyWith(

                          ),
                        ),
                      ),),
                      Flexible(flex: 1,child: Container(
                        alignment: Alignment.center,
                        child: Text(modelGoldSize.data!.sizes![index].japan!,
                          textDirection: TextDirection.ltr,
                          style: const TextStyle(fontSize: 12).copyWith(

                          ),
                        ),
                      ),),
                      Flexible(flex: 1,child: Container(
                        alignment: Alignment.center,
                        child: Text(modelGoldSize.data!.sizes![index].europe!,
                          textDirection: TextDirection.ltr,
                          style: const TextStyle(fontSize: 12).copyWith(

                          ),
                        ),
                      ),),
                    ],
                  ),
                ),
              ),
            );
          },
          ))
        ],
      ),
    );
  }


  calculateCircumference(value){
    circumference = value;
    print("Info calculateCircumference : $value");
  }
  calculateRadius(value){
    // ignore: avoid_print
    print("Info calculateRadius : $value");
  }
  calculateDiameter(value){
    diameter = value;
    print("Info calculateDiameter : $value");
  }

}
